File: static/js/classifiers.js
var GOOD = 1;
var NEUTRAL = 2;
var BAD = 3;
var UNCERTAIN = 4;

rivets.binders.coding = function(el, value) {
	var classToAdd;
	switch(value) {
		case GOOD:
			classToAdd = "label-success";
			break;
		case NEUTRAL:
			classToAdd = "label-warning";
			break;
		case BAD:
			classToAdd = "label-danger";
			break;
		default:
			classToAdd = "label-default";
	}
	$(el).addClass(classToAdd);
}

// Classifier generator for color coding
function makeSimpleClassifier(higherIsBetter, warnCount, okCount)
{
	var classifier;
	if(higherIsBetter) {
		classifier = function(count) {
			if(count > okCount) {
				return GOOD;
			} else if(count > warnCount) {
				return NEUTRAL;
			} else {
				return BAD;
			}
		}
	} else {
		classifier = function(count) {
			if(count < okCount) {
				return GOOD;
			} else if(count < warnCount) {
				return NEUTRAL;
			} else {
				return BAD;
			}
		}
	}
	return classifier;
}

// Classifier instance functions
var starClassifier = makeSimpleClassifier(true, 500, 1000);
var forkClassifier = makeSimpleClassifier(true, 150, 400);
var sizeClassifier = makeSimpleClassifier(false, 500000, 200000);
var dateClassifier = makeSimpleClassifier(false, 60, 30);

// Language classifier
function langClassifier(language) {
	switch (language) {
		case "C":
		case "C++":
		case "Objective-C":
			return BAD;
		case "Shell":
		case "Java":
		case "C#":
		case "Scala":
		case "Go":
			return NEUTRAL;
		case "JavaScript":
		case "Ruby":
		case "PHP":
		case "Python":
		case "Dart":
		case "Lua":
			return GOOD;
		default:
			return UNCERTAIN;
	}
}

File: classifiers.py
import datetime

# Define classification constants
GOOD = 1
NEUTRAL = 2
BAD = 3
UNCERTAIN = 4

# Creates simple classifiers for star, fork, size, date properties
def makeSimpleClassifier(higherIsBetter, neutralCount, goodCount):
	if higherIsBetter:
		def classifier(value):
			if value > goodCount:
				return GOOD
			elif value > neutralCount:
				return NEUTRAL
			else:
				return BAD
		return classifier
	else:
		def classifier(value):
			if value < goodCount:
				return GOOD
			elif value < neutralCount:
				return NEUTRAL
			else:
				return BAD
		return classifier

# Setup classifier functions
star = makeSimpleClassifier(True, 500, 1000)
fork = makeSimpleClassifier(True, 150, 400)
size = makeSimpleClassifier(False, 500000, 200000)
date = makeSimpleClassifier(False, 60, 30)

def lang(language):
	if language in ['C', 'C++', 'Objective-C']:
		return BAD
	elif language in ['Shell', 'Java', 'C#', 'Scala', 'Go']:
		return NEUTRAL
	elif language in ['JavaScript', 'Ruby', 'PHP', 'Python', 'Dart', 'Lua']:
		return GOOD
	return UNCERTAIN

def classify(repo):
	daysSinceLastUpdate = (datetime.datetime.utcnow() - repo.updated_at).days
	return {
		'stars': star(repo.stargazers_count),
		'forks': fork(repo.forks_count),
		'language': lang(repo.language),
		'size': size(repo.size),
		'date': date(daysSinceLastUpdate)
	}